skbio/metadata/_interval.py

```diff
@@ -0,0 +1,274 @@
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+                                  boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+                                  boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval
+        2. Query by key/val pair (i.e. gene=sagA)
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = []
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs += self._query_interval(value)
+
+        # Find queries by feature attribute
+        if metadata is not None:
+            for (key, value) in metadata.items():
+                invs += self._query_attribute(key, value)
+        return invs
+
+    def drop(self, intervals=None, boundaries=None, metadata=None):
+        """ Drops Interval objects according to a specified query.
+
+        Drops all Interval objects that matches the query.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+                                  boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+                                  boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.drop(metadata={'name': 'sagA'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        []
+        """
+        if intervals is None:
+            intervals = []
+        if metadata is None:
+            metadata = {}
+
+        queried_invs = self.query(intervals=intervals,
+                                  boundaries=boundaries,
+                                  metadata=metadata)
+        new_invs = []
+        # iterate through queries and drop them
+        for inv in self._metadata:
+            if inv not in queried_invs:
+                new_invs.append(inv)
+        self._metadata = new_invs
+        self._is_stale_tree = True
+
+    def __eq__(self, other):
+        return sorted(self._metadata) == sorted(other._metadata)
```

Comment:
should we? The interval tree is really just a copy of the `self._metadata`, restructured for faster querying.